### PR TITLE
App access: rewrite redirects to public app address from leaf cluster.

### LINKED ISSUE #8112: App access issues redirect to app's leaf cluster public addr
### Description

**What happened**:

User has a root and a leaf clusters and configured a Kibana application in the leaf cluster:

```
app_service:
  enabled: yes
  debug_app: true
  apps:
  - name: kibana
    uri: 'http://127.0.0.1:5601/kibana/app/home'
```

Trying to access the application from the root cluster's UI, a redirect occurs to `https://<leaf-cluster-public-addr>/kibana/app/home` which is not accessible since the leaf cluster is not exposed over public network.

**What you expected to happen**:

User should get redirected to the app's public address in root instead of leaf.

**Analysis**:

This is an issue with how Teleport does redirect in the scenario where an application URI has a path in it. Specifically, this piece of code appears to cause it: https://github.com/gravitational/teleport/blob/v7.1.0/lib/srv/app/transport.go#L130-L148.

If the app’s URI has a path in it, Teleport checks whether the request comes to the root path (`/`) and if so, does a redirect to the app’s path. The problem is that this redirect is happening on the leaf cluster in this case (specifically, inside the application access agent) so the redirect location gets rewritten with the app’s public address (which is why you’re seeing redirect to leaf public addr).

Instead, this redirect should be happening on the root cluster side (inside root’s proxy).

**Workarounds**:

- Specify app URI without the path, e.g. `http://localhost:5601` if the app is available on that address.
- Instead of using a trusted clusters, connect an application agent to the “root” cluster directly.

### Server Details
- Teleport version (run `teleport version`): All with app access

### PR DESCRIPTION
**Pull request by Tener**
Apps on the leaf cluster will occasionally issue redirects to their public address (see #8112). This is not what we want when the client is accessing the app via the root cluster: the redirect should be made relative to the root cluster's public address.

Fixes #8112.

### CODE REVIEW

**Review by jentfoo — COMMENTED** [review]
I am concerned that this strategy may not be safe.

`X-Forwarded-Host` is almost always appended to by proxies (rather than replaced).  This allows external users to manipulate this field pretty easily.  We had a former security issue around potentially obscuring your source in the audit record due to this: https://github.com/gravitational/teleport.e/issues/325

I recognize the use here is different, and that it's being used in a redirect.  But I still question if this is the best way to do this.  Have you considered:
* Can this value be configured rather than trying to parse from the request?
* I assume the root knows this information since it must communicate to the leaf, can instead the root securely communicate the leaf it's public ip?
* If this truly is the best way to do this, should it be enabled by default?  I feel like this should be disabled by default and only enabled when strictly required.

Let me know your thoughts around potential external manipulation and the other strategies mentioned above.  Thank you!

**Review by r0mant — CHANGES_REQUESTED** [review]
I agree with @jentfoo here, this solution seems hacky tbh. Can we find a way to fix it without relying on header overrides and introducing extra labels?

As suggested in the ticket's "analysis" section, can we handle this redirect on the proxy side instead?

**Comment by Tener** [comment]
@jentfoo 

> * Can this value be configured rather than trying to parse from the request?
If you want a static value to be used then you can already do that: the app's public address is the way to do it. The problem is that this doesn't scale: you need to configure it for every app in the leaf cluster, and it simply doesn't work if there are two root clusters (perhaps not a common setup, but possible). 

> * I assume the root knows this information since it must communicate to the leaf, can instead the root securely communicate the leaf it's public IP?
I think we might be able to check the certs provided with the connection and only use the `X-Forwarded-Host` if the connection was made from the root cluster.

> * If this truly is the best way to do this, should it be enabled by default?  I feel like this should be disabled by default and only enabled when strictly required.
I discussed this with @smallinsky, and enabled-by-default was a result of that discussion, but if there are security concerns then I concur this should be opt-in instead.

@r0mant 

> I agree with @jentfoo here, this solution seems hacky tbh. Can we find a way to fix it without relying on header overrides and introducing extra labels?

This will need a feature flag either way, as the current behavior may be what some people want. 
 
> As suggested in the ticket's "analysis" section, can we handle this redirect on the proxy side instead?

I will look into fixing this on the proxy side.

**Comment by r0mant** [comment]
@Tener @smallinsky 

> This will need a feature flag either way, as the current behavior may be what some people want.

I'm not sure why this needs a flag? The current behavior is that app access doesn't work for leaf cluster's apps (in that particular scenario when you use a path). I don't think people want this behavior? Can we just make it work in all scenarios without any extra flags?

**Comment by smallinsky** [comment]
https://github.com/gravitational/teleport/pull/21067#issuecomment-1413952539

Agree, The only reason to have  switch to disable this behaviour is where this fix will be causing unexpected side effect behaviour.

**Comment by Tener** [comment]
@r0mant @smallinsky
On further thought, I agree; while the current behavior is not _entirely_ broken (i.e., the clients may have access to the leaf cluster under some scenarios), this isn't really what we want for root clusters. No feature flags, then.

**Comment by Tener** [comment]
@smallinsky @r0mant @jentfoo PTAL

**Review by jentfoo — APPROVED** [review]
This strategy looks much better to me :+1:

**Review by mdwn — APPROVED** [review]
This seems fine to me!

File: lib/web/app/transport.go [hunk 1]
```diff
@@ -139,9 +145,69 @@ func (t *transport) RoundTrip(r *http.Request) (*http.Response, error) {
...
+}
+
+// rewriteRedirect rewrites redirect to a public address of downstream app.
+func (t *transport) rewriteRedirect(resp *http.Response) error {
+	if !t.c.remoteApp() {
+		// Not a downstream app, short circuit.
+		return nil
+	}
+
+	if !isRedirect(resp.StatusCode) {
```
[the remark this answers is not in the document]
Comment (Tener) [reply lib/web/app/transport.go:172]: Function moved to utils and reused in both spots.

File: lib/web/app/transport.go [hunk 2]
```diff
@@ -78,6 +79,11 @@ func (c *transportConfig) Check() error {
 	return nil
 }
 
+// remoteApp returns true if the route to app points at remote cluster.
+func (c *transportConfig) remoteApp() bool {
```
[the remark this answers is not in the document]
Comment (Tener) [reply lib/web/app/transport.go]: Done.

File: lib/web/app/transport.go [hunk 3]
```diff
@@ -139,9 +145,69 @@ func (t *transport) RoundTrip(r *http.Request) (*http.Response, error) {
 		return nil, trace.Wrap(err)
 	}
 
+	if err = t.rewriteRedirect(resp); err != nil {
```
[the remark this answers is not in the document]
Comment (Tener) [reply lib/web/app/transport.go:151]: Sure.

File: lib/web/app/transport.go [hunk 4]
```diff
@@ -139,9 +145,69 @@ func (t *transport) RoundTrip(r *http.Request) (*http.Response, error) {
...
+	t.servers.Range(func(serverID, appServerInterface interface{}) bool {
+		appServer, ok := appServerInterface.(types.AppServer)
+		if !ok {
+			t.c.log.Warnf("Failed to load AppServer, invalid type %T", appServerInterface)
+			return true
+		}
+
+		// If this is a redirect to a public addr of a leaf cluster, rewrite it.
+		// We want the rewrite to happen using our own public address.
+		if host == appServer.GetApp().GetPublicAddr() {
```
[the remark this answers is not in the document]
Comment (Tener) [reply lib/web/app/transport.go]: Indeed, thanks, this simplifies the check quite a bit.

File: lib/web/app/transport.go [hunk 5]
```diff
@@ -139,9 +145,69 @@ func (t *transport) RoundTrip(r *http.Request) (*http.Response, error) {
...
+			// reset the scheme and host, leaving path alone.
+			// this will turn the redirect into relative one.
+			u.Scheme = ""
+			u.Host = ""
```
[the remark this answers is not in the document]
Comment (Tener) [reply lib/web/app/transport.go]: Good call about returning `u.Path`, this makes the code cleaner. You are also right that an empty location can appear, I believe we should make the location "/" in such a case.

**Comment by Tener** [comment]
@r0mant PTAL.

File: lib/web/app/transport.go [hunk 6]
```diff
@@ -139,9 +145,58 @@ func (t *transport) RoundTrip(r *http.Request) (*http.Response, error) {
...
+	}
+
+	host, _, err := net.SplitHostPort(u.Host)
+	if err != nil {
+		host = u.Host
+	}
+
+	// If this is a redirect to a public addr of a leaf cluster, rewrite it.
+	// We want the rewrite to happen using our own public address.
+	if host == t.c.identity.RouteToApp.PublicAddr {
```
Comment (r0mant) [inline lib/web/app/transport.go:188]: Is this check really needed? Wouldn't we want to rewrite redirect location to root's address in any case?

**Review by r0mant — APPROVED** [review]

File: lib/web/app/transport.go (hunk 6 above)
Comment (Tener) [reply lib/web/app/transport.go:188]: Redirects to other addresses - like `https://goteleport.com/` - should stay as they are.

**Review by smallinsky — APPROVED** [review]

**Pull request merged**